Fixtures: Modularize fixtures creating AiiDA test instance and profile

sphuber opened the pull request:
Fixes #5526 
Fixes #4365 

The `aiida_profile` is a fixture that will automatically load an
existing test profile, if specified by the `AIIDA_TEST_PROFILE` environment
variable, or otherwise create a temporary and isolated test profile from
scratch. While very useful to make it easy to start writing tests
against an AiiDA instance, it hard-coded most if not all of the
configuration.

For example, the storage backend was hardcoded to the `core.psql_dos`
storage backend. This used to be fine since this was the only storage
backend available, but since `aiida-core==2.1` they are pluginnable and
so it becomes necessary to be able to configure them.

The `aiida_profile` still works the same as in that it automatically
provides a fully loaded AiiDA test instance, and it still loads the
`core.psql_dos` storage by default, but it does so through the new
`aiida_profile_factory` fixture. This fixture can be reused in plugin
packages to achieve the same effect, but they can override any setting
of the profile configuration by providing those through the
`custom_configuration` argument.

These new fixtures make the `aiida.manage.tests.main` module obsolete.
It was added when the tests were still run with the `unittest` module
instead of `pytest`. The module is deprecated as all of its
functionality is completely provided by the new, more modular and more
succinct pytest fixtures.

sphuber commented:
I am successfully using these fixtures in [`aiida-s3`](https://github.com/sphuber/aiida-s3/) to run with minimal changes against a custom storage backend.

ltalirz commented on aiida/manage/tests/pytest_fixtures.py:
```diff
@@ -42,27 +61,243 @@ def aiida_caplog(caplog):
-def aiida_profile():
-    """Set up AiiDA test profile for the duration of the tests.
+@pytest.fixture(scope='session')
+def postgres_cluster(
```
it looks to me like some of the code below duplicates what is already in the `TemporaryProfileManage` in `main.py`. Can we avoid that?

sphuber replied:
Not in a way that I can see that will make things more clear. The problem of the `TemporaryProfileManager` is that it doesn't properly separate concerns but mixes various concepts: The concept of a profile, but hard-codes the integration with a disk repository and a PostgreSQL cluster for a database. The whole point is to properly decouple this so that parts can be easily replaced where needed. I think it is better to deprecate the entire module as I have done, then try and keep bits around and make things complicated. I would have even removed it entirely instead of deprecating it, because the main use case is for plugin to use `aiida_profile`, but I am not sure if code out there used these classes directly.

ltalirz reviewed (commented):
Cheers, @sphuber 

The main downside I see is that it appears that this change will break the tests in pretty much all AiiDA plugins, which I consider to be quite serious (even if it is easy to fix).

Is there no way to "disable" the fixture for specific tests where it is not needed?

sphuber commented:
> Is there no way to "disable" the fixture for specific tests where it is not needed?

That's the problem, since it is `autouse=True` and session-scoped, it will be loaded as soon as the tests run. This means everything is created and loaded instantly and for example in `aiida-s3` I don't even get the chance to load another type of profile.

sphuber commented:
Ok, I think I may have found a way, but it will require some refactoring of the `PsqlDosBackend` and the `PsqlDostoreMigrator`. Need to clean things up though, but will do that later. In short, we should put this on hold until I have cleaned things up.

sphuber commented:
@ltalirz with #5760 merged, I have rebased this PR and it is now ready for review. The `aiida_profile` fixture remains `autouse=True`.

ltalirz reviewed (commented):
thanks @sphuber !

I assume the relevant code is copied from the `aiida.manage.tests`, so I did not check all the logic down to the last detail.
If there is some change implementation that you'd like me to look at in particular, let me know.

Before I approve, could you please have a look at `conftest.py`? It seems to contain a number of fixtures that are very similar to the ones now present in the `pytest_fixtures.py` (profile_factory, isolated_config, ...)

sphuber commented on .github/system_tests/pytest/test_memory_leaks.py:
```diff
@@ -23,6 +25,7 @@ def run_finished_ok(*args, **kwargs):
     assert node.is_finished_ok, (node.exit_status, node.exit_message)
 
 
+@pytest.mark.usefixtures('aiida_profile')
```
I've removed them, but I am not sure that it will work. These tests are called in a different way, outside of the normal test suite. They don't have the `conftest.py` which imports the `aiida.manage.tests.pytest_fixtures` which is responsible for loading the `aiida_profile` fixture.

sphuber replied:
They did run, but that is only because we explicitly pass the `conftest.py` of the `tests` folder. I think these tests were placed outside of the suite originally when we were still half running `unittest` in the migration to `pytest`. I will move them to the `tests` folder now, since they don't need to be separated any longer.

sphuber commented:
> I assume the relevant code is copied from the `aiida.manage.tests`, so I did not check all the logic down to the last detail. If there is some change implementation that you'd like me to look at in particular, let me know.

Actually, the idea is the same, but I didn't salvage much of the code, or at the least it has been rewritten for a large part. That being said, I have tested them against the `aiida-s3` tests, so reasonably confident that they work out of the box.

> Before I approve, could you please have a look at `conftest.py`? It seems to contain a number of fixtures that are very similar to the ones now present in the `pytest_fixtures.py` (profile_factory, isolated_config, ...).
> It seems to me we may be able to get rid of those?

These are actually quite different. They only provide a simple AiiDA instance (`.aiida` folder with `config.json`) and are used internally just to test functionality that operates on the config file: setting options, adding/removing profiles etc. This is not something that plugin packages should have to do. And the config/profiles that are being generated are "inert" as in they don't configure a working storage backend with the required services, so it couldn't be used for running tests anyway. There may be a slight bit of overlap that can be reduced, but that would only be for internal use, which I can take care of later.

> The rationale for placing a fixture into `conftest.py` vs. the `pytest_fixtures.py` is that the former is only for aiida-core, while fixtures in the latter become available to all plugins, and should thus be considered public API.

Exactly right.

> Edit: perhaps we should mention this rationale in the module docstrings of these two modules.

I have updated the module docstrings accordingly

Finally, in the added commit I have improved/corrected the docstrings of the fixtures and added them to the fixture documentation with instructions and examples on how to use them.

ltalirz commented:
Thanks! 

In that case I should still have a quick look at the logic and check whether there are any changes with respect to the one in the previous test manager (tomorrow).

Also, I noticed that the system tests are still using the test manager module, and so there are a couple of `AiidaDeprecationWarning`s in the test logs. Perhaps it makes sense to move these over?

sphuber commented:
> Also, I noticed that the system tests are still using the test manager module, and so there are a couple of AiidaDeprecationWarnings in the test logs. Perhaps it makes sense to move these over?

These are actually testing the test manager module itself. Since we are just deprecating it and not removing them yet, it seems best to keep the tests around as well.

sphuber commented on aiida/manage/tests/pytest_fixtures.py:
```diff
@@ -42,16 +69,260 @@ def aiida_caplog(caplog):
+def postgres_cluster(
+    database_name: str | None = None,
+    database_username: str | None = None,
+    database_password: str | None = None
```
shall I just get rid of it entirely?

sphuber replied:
The `Postgres` class still requires passing an explicit password. Is the idea that when using pgtest that *any* password will do? So should I just remove the argument from `postgres_cluster` and pass a random password to `Postgres.create_dbuser`?

ltalirz replied:
> The Postgres class still requires passing an explicit password.

I see.

>  Is the idea that when using pgtest that any password will do? So should I just remove the argument from postgres_cluster and pass a random password to Postgres.create_dbuser?

I would guess that even an empty password should do

sphuber replied:
Ok. But does it even make sense to expose the other variables? Would a user ever want to specify an explicit username? If so, then the same could hold for the database, couldn't it? So either we maybe remove both and just keep the database name or we keep all three. But maybe the database name should also not be configurable if the whole point is that it should just create a temporary one. Not sure if you would ever need to have to specify an explicit one

sphuber commented on aiida/manage/tests/pytest_fixtures.py:
```diff
@@ -42,16 +69,260 @@ def aiida_caplog(caplog):
+        """
+        config = aiida_instance
+        configuration = {
+            'default_user_email': 'test@aiida.local',
```
Can do. Not sure it matters though, because it people want to override it, they pass it in the `custom_configuration`. They wouldn't have to start messing around with a module constant.

sphuber commented on aiida/manage/tests/pytest_fixtures.py:
```diff
@@ -42,16 +69,260 @@ def aiida_caplog(caplog):
+            yield configuration.CONFIG
+        finally:
+            if reset:
+                if current_path_variable:
```
Yeah, I think you are right and I agree that restoring to original state, even if that means unsetting, is the correct behavior
